sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -1571,6 +1659,162 @@ a normal SYCL command-group submission.
 associated with the graph node resulting from this command-group submission is
 different from the one with which the dynamic_parameter was created.
 
+|
+[source,c++]
+----
+void handler::ext_oneapi_barrier(const std::vector<dynamic_event>& waitList);
+----
+
+|Overload of `ext_oneapi_barrier` that takes a list of dynamic events to wait
+on.
+
+This function has the same semantics as `ext_oneapi_barrier(const
+std::vector<event>&)`.
+
+|
+[source,c++]
+----
+void handler::depends_on(dynamic_event depEvent);
+----
+
+|Overload of `depends_on()` which takes a dynamic event dependency.
+
+This function has the same semantics as `depends_on(event)`.
+
+Parameters:
+
+* `depEvents` - Dynamic event dependency.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if this function is called from
+a normal SYCL command-group submission.
```

Comment:
Reviving an idea that we discussed a while ago ... We could design the API to enforce these restrictions through the type system.  To do this, we'd change `command_graph::add` so that the `cgf` callback does *not* take a `sycl::handler`.  Instead, we'd pass some different object to the callback like `sycl::ext::oneapi::experimental::add_handler`.  The `add_handler` would support most of the same member functions as `handler`, omitting those that cannot be used when constructing a graph.  Member functions that are specific to graphs (like this overload of `ext_oneapi_barrier` taking a vector of `dynamic_event`) would be added only to `add_handler`.  Thus, it would be impossible to call these functions from a "normal" SYCL command-group function.

Reply:
I'm not very keen on this idea because it excludes these features from recording APIs and creates unituitive differences between command groups in either API.

An alternative could be that we allow the `dynamic_event` overloads of these functions in normal SYCL code and just have it use whatever the current event value of the `dynamic_event` is.

Reply:
> I'm not very keen on this idea because it excludes these features from recording APIs and creates unituitive differences between command groups in either API.

Do you really think people will use recording mode and the explicit APIs at the same time like this?  It seems like users who are willing to change their code would just use the explicit API.  Those users who don't want to change their code will use the recording mode.

> An alternative could be that we allow the `dynamic_event` overloads of these functions in normal SYCL code and just have it use whatever the current event value of the `dynamic_event` is.

This wouldn't really solve the problem because a `dynamic_event` can be constructed with no underlying SYCL event.  So, you'd still need special checking in that case.

Reply:
> Do you really think people will use recording mode and the explicit APIs at the same time like this?

I don't think simultaneous use of the APIs comes into this, and ignores the main point which is creating a feature disparity between the APIs is very undesirable. The fact that it is unintuitive and also creates additional spec bulk is secondary.

> Those users who don't want to change their code will use the recording mode.

I also don't think you can put API preference into such strict terms as this. Recording style APIs are very popular in existing applications using other existing graph APIs, likely because it involves very little change to the way you write your general SYCL/CUDA/whatever code. Using more advanced features like this will likely always require some amount of modification, and that does not take away from the simple use case where you can wrap your code in a graph without modification, but it also doesn't mean that usage of the recording APIs should be limited to that most simple of use cases only.

> This wouldn't really solve the problem because a dynamic_event can be constructed with no underlying SYCL event. So, you'd still need special checking in that case.

Attempting to use a `dynamic_event` in a graph that has no underlying event is [already an error](https://github.com/reble/llvm/blob/9fe7edfa9c1cb6cd6fbc63c6b62a8f24d8339cbf/sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc?plain=1#L1778) (though that wording is missing from the definitions of the new overloads taking dynamic events, I'll fix that) so this would be an error in both cases if the wording is changed to also include this use.